src/common/scaled-scene-buffer.c

```diff
@@ -176,7 +232,9 @@ scaled_scene_buffer_create(struct wlr_scene_tree *parent,
 	self->impl = impl;
 	self->active_scale = 1;
 	self->drop_buffer = drop_buffer;
+	self->cached_buffers = cached_buffers;
 	wl_list_init(&self->cache);
+	wl_list_insert(self->cached_buffers, &self->link);
```

Comment:
I think we should either `assert(cached_buffers)` or check for `self->cached_buffers` in `_update_buffer()`. Personally I prefer the 2nd variant as that allows future implementations that don't want to get their things cached (e.g. they pass `NULL` as arg and don't provide `impl->equal()`).

Reply:
Regarding the `wl_list_remove(&self->link)` in `_handle_node_destroy()`: that can be solved by either checking for `self->cached_buffers` before calling or by `wl_list_init(&self->link)` in `_create()` on `!cached_buffers`.

Or with an alternative: a pattern often used in wayland protocol implementations within wlroots. E.g. if you have a `struct wl_list` which is either part of some list or not it can always be removed via `wl_list_remove()` if it has been initialized after removal.

Reply:
I agree. Added checks for `self->cached_buffers` in `scaled_scene_buffer_create()` and `_update_buffer()`.

Reply:
`wl_list_remove(&self->link);` in `_handle_node_destroy` will crash. See above

Reply:
Sorry, I missed your second comment. I copied your suggested code.

Reply:
I thought so thus the 2nd reply :)